### PR TITLE
Make a temporary array static to exclude extra allocations

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

The `map` array is allocated every the method call and never changed. So replace it with `if` statements to exclude extra allocations.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/lang/parserutils.cs [hunk 1]
```diff
@@ -550,6 +541,15 @@ private static RegexOptions parseRegexOptions(SplitOptions options)
             return result;
         }
 
+        private static readonly int[][] map = {
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/lang/parserutils.cs]: While not as pretty, I'd expect removing the arrays and using 6 if statements would result in faster code.
With the arrays, there are between 6 and 12 array bounds checks in addition to the 6 tests.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/lang/parserutils.cs]: How about using `Tuple<SplitOptions, RegexOptions>[]` instead? So we can keep it pretty and also a bit more performant.

Comment (iSazonov) [reply src/System.Management.Automation/engine/lang/parserutils.cs]: It is again an array and a cycle - seems `if` will be more faster.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/lang/parserutils.cs]: JIT eliminates the range check when looping on an array as IEnumerable, so the range check only happens to `entry[0]` and `entry[1]` in the original code. By replacing with a tuple, we can avoid the range change. But I'm fine changing it to if's too.

**Comment by iSazonov** [comment]
@daxian-dbw I updated the PR summary.

**Review by lzybkr — APPROVED** [review]

**Pull request merged**